File: core/src/main/java/moe/caa/multilogin/core/auth/yggdrasil/YggdrasilAuthenticationResult.java
package moe.caa.multilogin.core.auth.yggdrasil;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.ToString;
import moe.caa.multilogin.api.auth.GameProfile;
import moe.caa.multilogin.core.configuration.yggdrasil.YggdrasilServiceConfig;

/**
 * HasJoined 验证结果
 */
@Getter
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@ToString
public class YggdrasilAuthenticationResult {
    private final Reason reason;
    private final GameProfile response;
    private final int yggdrasilId;
    private final YggdrasilServiceConfig yggdrasilServiceConfig;

    /**
     * 提供一个允许登录的结果
     */
    protected static YggdrasilAuthenticationResult ofAllowed(GameProfile response, int yggdrasilId, YggdrasilServiceConfig yggdrasilServiceConfig) {
        return new YggdrasilAuthenticationResult(Reason.ALLOWED, response, yggdrasilId, yggdrasilServiceConfig);
    }

    /**
     * 提供一个服务器宕机的结果
     */
    protected static YggdrasilAuthenticationResult ofServerBreakdown() {
        return new YggdrasilAuthenticationResult(Reason.SERVER_BREAKDOWN, null, -1, null);
    }

    /**
     * 提供一个验证失败的结果
     */
    protected static YggdrasilAuthenticationResult ofValidationFailed() {
        return new YggdrasilAuthenticationResult(Reason.VALIDATION_FAILED, null, -1, null);
    }

    /**
     * 提供一个没有验证服务器的结果
     */
    protected static YggdrasilAuthenticationResult ofNoService() {
        return new YggdrasilAuthenticationResult(Reason.NO_SERVICE, null, -1, null);
    }

    public enum Reason {
        // 通过
        ALLOWED,
        // 服务器宕机或破坏
        SERVER_BREAKDOWN,
        // 验证失败
        VALIDATION_FAILED,
        // 没有验证服务器
        NO_SERVICE;
    }
}

File: velocity/src/main/kotlin/moe/caa/multilogin/velocity/auth/yggdrasil/YggdrasilAuthenticationResult.kt
package moe.caa.multilogin.velocity.auth.yggdrasil

import moe.caa.multilogin.api.profile.GameProfile
import moe.caa.multilogin.velocity.auth.validate.ValidateContext
import moe.caa.multilogin.velocity.config.service.yggdrasil.BaseYggdrasilService

/**
 * 表示 Yggdrasil 验证结果返回
 */
sealed interface YggdrasilAuthenticationResult {

    /**
     * Yggdrasil 验证成功返回
     */
    data class Success(
        val baseYggdrasilService: BaseYggdrasilService, val profile: GameProfile
    ) : YggdrasilAuthenticationResult {
        fun buildValidateContext() = ValidateContext(baseYggdrasilService, profile)
    }

    /**
     * 验证失败返回
     */
    data class Failure(val reason: Reason) : YggdrasilAuthenticationResult {

        // 从上往下, ordinal 越大 表示越严重
        enum class Reason {
            // 没有找到任何有效的 Yggdrasil Service
            NO_YGGDRASIL_SERVICES,
            // 无效的会话
            INVALID_SESSION,
            // 服务器无法连接
            SERVER_BREAK_DOWN,
        }
    }
}

